helm_chart/hookexecutor/pre_delete_test.go

```diff
@@ -0,0 +1,316 @@
+    app: ubiquity-db
+    product: ibm-storage-enabler-for-containers
+  name: ubiquity-db-868c4cb89d-b2xvs
+  namespace: ubiquity
+spec:
+  containers:
+  - image: ibmcom/ibm-storage-enabler-for-containers-db:1.2.0
+    name: ubiquity-db
+    volumeMounts:
+    - mountPath: /var/lib/postgresql/data
+      name: ibm-ubiquity-db
+      subPath: ibm-ubiquity
+    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
+      name: default-token-lrwfk
+      readOnly: true
+  volumes:
+  - name: ibm-ubiquity-db
+    persistentVolumeClaim:
+      claimName: ibm-ubiquity-db
+  - name: default-token-lrwfk
+    secret:
+      defaultMode: 420
+      secretName: default-token-lrwfk
+`
+
+var test_one = int32(1)
+var test_zero = int32(0)
+
+var _ = Describe("PreDelete", func() {
+
+	var e Executor
+	var kubeClient *fakekubeclientset.Clientset
+	var pvc *v1.PersistentVolumeClaim
+	var pv *v1.PersistentVolume
+	var deploy *appsv1.Deployment
+	var pod *v1.Pod
+
+	BeforeEach(func() {
+
+		pvcObj, _ := FromYaml([]byte(test_pvcYaml))
+		pvc = pvcObj.(*v1.PersistentVolumeClaim)
+
+		pvObj, _ := FromYaml([]byte(test_pvYaml))
+		pv = pvObj.(*v1.PersistentVolume)
+
+		deployObj, _ := FromYaml([]byte(test_deployYaml))
+		deploy = deployObj.(*appsv1.Deployment)
+
+		podObj, _ := FromYaml([]byte(test_podYaml))
+		pod = podObj.(*v1.Pod)
+
+		os.Setenv("UBIQUITY_DB_PV_NAME", pv.Name)
+		os.Setenv("NAMESPACE", "ubiquity")
+
+		kubeClient = fakekubeclientset.NewSimpleClientset(pvc, pv, deploy, pod)
+
+		e = PreDeleteExecutor(kubeClient)
+	})
+
+	AfterEach(func() {
+		os.Setenv("UBIQUITY_DB_PV_NAME", "")
+		os.Setenv("NAMESPACE", "")
+	})
+
+	Describe("test Execute", func() {
+
+		Context("delete UbiquityDB Pods", func() {
+
+			BeforeEach(func() {
+				deploy, err := kubeClient.AppsV1().Deployments(deploy.Namespace).Get(deploy.Name, metav1.GetOptions{})
+				Ω(err).ShouldNot(HaveOccurred())
+				Expect(deploy.Spec.Replicas).To(Equal(&test_one))
+
+				go func() {
+					podWatcher := watch.NewFake()
+					kubeClient.PrependWatchReactor("pods", testcore.DefaultWatchReactor(podWatcher, nil))
+
+					time.Sleep(40 * time.Millisecond)
+					podWatcher.Delete(pod)
+				}()
+			})
+
+			It("should be deleted successfully by setting replicas to 0", func(done Done) {
+				err := e.(*preDeleteExecutor).deleteUbiquityDBPods()
+				Ω(err).ShouldNot(HaveOccurred())
+
+				deploy, err := kubeClient.AppsV1().Deployments(deploy.Namespace).Get(deploy.Name, metav1.GetOptions{})
+				Ω(err).ShouldNot(HaveOccurred())
+
+				// we only check if the replicas is updated since the fake server won't trigger any pod changes.
+				Expect(deploy.Spec.Replicas).To(Equal(&test_zero))
+
+				close(done)
+			})
+		})
+
+		Context("delete UbiquityDB Pods when Deployment is gone", func() {
+
+			BeforeEach(func() {
+				err := kubeClient.AppsV1().Deployments(deploy.Namespace).Delete(deploy.Name, nil)
+				Ω(err).ShouldNot(HaveOccurred())
+			})
+
+			It("should return without error", func(done Done) {
```

Comment:
Just change the IT test to : should succeed even if ubiqutiydb deployment not exist (idempotancy).